### ISSUE TITLE
Unified media handling

### ISSUE THREAD

**Issue by eyelidlessness** [enketo/enketo-express#464]
In #459 I discussed some future steps we could take to further simplify media URL logic. This issue will be used to track those changes. Depending on the size, and coherent overlap of the work described here, I may submit a single PR, or break it into smaller chunks.

## Tasks

- [ ] **Generate media URLs client-side:** there's little remaining barrier to moving media URL parameterization to the client. Besides reducing unnecessary scope of the server's responsibilities, it will also unblock some of the other changes discussed in the first PR.

- [ ] **Offline-capable requests for media:** currently these requests are handled by replacing `src` attributes with `data-offline-src`, with an additional step to look those up and initiate media loading each time a form/instance is loaded, when a form is reset, when repeat instances are added, and so on. This distinction can be eliminated, instead handling those requests in `fetch` events within the Service Worker. We may also be able to handle caching at that level rather than in IndexedDB, but I'll likely split that off as a separate work item and defer it for another future follow-up.

- [ ] **Consistent handling of media URL resolution and requests:** there are several different ways URLs are resolved, depending on their type/role in the form: binary defaults, media in labels, external secondary instances, last-saved, instance attachments. I plan to unify these so that there is one consistent way that all media URLs are resolved and requested/fulfilled.

It may be tempting to combine these last two tasks, as a Service Worker is capable of handling both and it would enable even further simplification. Unfortunately, Service Workers cannot run in some browsers' private/incognito modes. While this is already a restriction of offline-capable mode, it's not a reasonable limitation for online-only, edit or preview modes. In any case, consistent handling of URL resolution can be shared, and hopefully the offline portion will just be a small wrapper around that specific to the pertinent Service Worker APIs.

**Comment by eyelidlessness** [comment]
I started moving media URL parameterization to the client, but backed off for now because I'm not sure it's necessary for the other parts.

I've made pretty good progress on handling offline-capable requests in the Service Worker. For now, my work in progress:

- Is a mess, I'll clean up and push in the morning.
- Works without `data-offline-src`.
- _Always_ attempts to make the network request, falling back to the offline `Cache` response only as necessary. This will also address #398 without any change to hashing logic.

I've also explored part of consistent handling of request resolution, which is now handling last-saved the same way other external secondary instances are handled. I'm feeling good about the general approach, but there's much more left to do on that front.

**Comment by lognaturel** [comment]
> Always attempts to make the network request

Are there etag or last-modified headers on the resources to prevent duplicate downloads? If so, then this sounds great.

**Comment by eyelidlessness** [comment]
The requests are directly passed through from the browser and the responses are returned and stored in cache exactly as they’re received, so it’ll be the same request-response cycle online-only mode would have based on the browser’s own caching logic, falling back to whatever response was previously received in network failure cases.

The Service Worker `Cache` API is surprisingly well designed for this use case, it’s basically `Map<Request, Response>` with some async, because async happens.

**Comment by lognaturel** [comment]
I think that makes sense. One bad thing I can think of would be if the browser expires the files before Enketo would have previously. I don't know enough about how browser caching works to know whether that's a possibility.

While thinking through this I noticed that offline forms with 4.1.2 show a lot more errors in the network inspector than 4.1.1 did. For example, this seems to happen for every media file regardless of whether or not there is connectivity:
```
Error: NetworkError when attempting to fetch resource.
    getMediaFile connection.js:598
    promise callback*getMediaFile connection.js:597
    request3 form-cache.js:385
    promise callback*_loadMedia/< form-cache.js:382
    _loadMedia form-cache.js:377
    updateMedia form-cache.js:322
    promise callback* enketo-webform.js:47
    <anonymous> enketo-webform.js:58674
```

I also see that the offline enabled image (little icon on top left of screen) is not working:
`GET https://test.getodk.cloud/x/images/offline-enabled.png
[HTTP/1.1 404 Not Found 0ms]`

These can certainly be split off but wanted to mention I noticed them.

**Comment by eyelidlessness** [comment]
> One bad thing I can think of would be if the browser expires the files before Enketo would have previously.

The Cache API is explicitly separated from other cache eviction policies, and managed by the browser’s quota. The cache storage quota ratio only changes when online and attempting to cache resources that were accessed by network. IndexedDB uses the same quotas. According to the docs, *everything* gets invalidated when anything does.

I think it would be a good idea to evict unused resources sooner (whichever access time strategy makes the most sense), but also don't consider it a blocker because the existing logic doesn’t handle this case either.

I’d really like to know more about the network errors, aside from the offline icon. If there’s a regression they should help reveal it.

**Comment by eyelidlessness** [comment]
I’m supposed to be taking a break, but I had a realization earlier that I can’t get out of my brain now. I’m writing it up now so I can get on with a real break. The changes in #465 lay a foundation which I believe allows this additional task:

- [ ] **Eliminate `form-cache.js` entirely:** currently this doesn't work, but only because the transform request is a `POST`. But it doesn’t need to be, the body is just an optional `xformUrl` which could just as reasonably be a query parameter. It seems likely that the reasoning behind it being a `POST` is that transformation results are cached in Redis, but that’s really an implementation detail rather than a meaningful state change. The request really **should be** considered idempotent from a client perspective.

That should be a trivial change, and wouldn’t be breaking because the API endpoint is internal.

In the future we might also consider whether other _actually non-idempotent_ requests (e.g. submissions) could be cached in a more generalized backlog, rather than the more resource-specific design currently in IndexedDB. This would probably be a considerably bigger change, deserving a separate issue which I’ll probably file when we’re ready to close this out.

**Comment by lognaturel** [comment]
> The request really should be considered idempotent from a client perspective.

Not really important for this specific issue but a thought that's worth capturing: I have this same discomfort about [the public API](https://apidocs.enketo.org/v2). Each endpoint has `GET` and `POST` variants with `GET` providing a subset of what `POST` does. I can't think of any reason to have both. The only vague idea I came up with is that some clients might want to know if a survey doesn't exist (`GET`'s 404). But then it seems the only thing they'd want to do with that information is create the form so they would have been better off going with the `POST` initially.

**Comment by MartijnR** [comment]
> Each endpoint has GET and POST variants with GET providing a subset of what POST does. I can't think of any reason to have both.

I remember this one! :) The GET requests are only for developers to make a little easier to test (there is a line about that in the documentation mentioned under each GET endpoint). I believe I stole that idea from somewhere else. Stripe maybe.

**Comment by lognaturel** [comment]
> there is a line about that in the documentation mentioned under each GET endpoint

Oh! "The GET version of this endpoint is only meant to test the Enketo Smart Paper API in the browser. It does not launch new surveys and is therefore not of use for integrating with Enketo." Now that you mention it I think you and I talked about this a thousand years ago and I'd totally forgotten and also apparently have forgotten how to read. Thank you! In a parallel universe, the decision could have been made to always use `GET` for requests that implicitly result in a transformed form being cached. We're not in that world so can't change it for the public API but seems reasonable to change it for the private API being discussed here.
